[Unified PDF] Zoom not working in WKWebView

hortont424 opened the pull request:
#### a30c570b4fcbafa412fed2b96ec5fe45f46fb52f
<pre>
[Unified PDF] Zoom not working in WKWebView
<a href="https://bugs.webkit.org/show_bug.cgi?id=288186">https://bugs.webkit.org/show_bug.cgi?id=288186</a>
<a href="https://rdar.apple.com/145047812">rdar://145047812</a>

Reviewed by Abrar Rahman Protyasha.

By default, WebKit respects &quot;user-scalable=no&quot;. Safari overrides this to force
pages to always be scalable. For multiple different reasons, all fixed in this patch,
the default viewport configuration for main frame UnifiedPDFPlugin had &quot;user-scalable=no&quot; set.

* Source/WebCore/html/PluginDocument.cpp:
Don&apos;t explicitly set user-scalable=no...

* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::ViewportConfiguration::pluginDocumentParameters): Deleted.
Remove this PDFPlugin-specific code from WebCore.

* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::forceAlwaysUserScalableMaximumScale const):
(WebCore::ViewportConfiguration::forceAlwaysUserScalableMinimumScale const):
Don&apos;t let &quot;force always user-scalable&quot; to ever *reduce* the available scale range, only increase it.
This prevents it from reducing our PDF maximum scale from 6x to 5x.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::viewportParameters):
Move the UnifiedPDFPlugin default viewport parameters computation to a static method on UnifiedPDFPlugin.
Critically, change the maximum scale to match UnifiedPDFPlugin&apos;s normal maximum scale,
and set `allowsUserScaling` to true (e.g. &quot;user-scalable=yes&quot;). The combination
of these two things make scaling work when in the default &quot;do not force always user-scalable&quot; state.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::resetViewportDefaultConfiguration):
Ask UnifiedPDFPlugin for its viewport configuration instead of WebCore.

Canonical link: <a href="https://commits.webkit.org/290779@main">https://commits.webkit.org/290779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccf33cb27265a3a98680c9549a3528c21e9cb427

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91113 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10656 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/150 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/96117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/41875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93163 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11062 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18974 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/96117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/41875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94114 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/82557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/96117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/126 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41008 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/143 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98098 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18315 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18574 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78385 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78232 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19331 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22739 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/94 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18318 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18042 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21503 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19819 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/html/PluginDocument.cpp:
```diff
@@ -107,11 +107,6 @@ void PluginDocumentParser::createDocumentStructure()
 
-#if PLATFORM(IOS_FAMILY)
-    // Should not be able to zoom into standalone plug-in documents.
-    document.processViewport("user-scalable=no"_s, ViewportArguments::Type::PluginDocument);
```
I think this is the only use of `Type::PluginDocument`. Maybe we can want to get rid of this case. We don't base any logic on this case anyways it seems

hortont424 replied:
ooh, looking

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -4490,6 +4490,22 @@ static bool isEmpty(PDFSelection *selection)
+    parameters.widthIsSet = true;
+    parameters.allowsUserScaling = true;
+    parameters.allowsShrinkToFit = false;
+    parameters.minimumScale = 1;
```
Lest we ever change this?
```suggestion
    parameters.minimumScale = minimumZoomScale;
```

hortont424 replied:
isn't it 0.2? I think that would be a significant behavior change, which I did not want to make in this patch

aprotyas replied:
oh, that's fair

aprotyas replied:
Actually, should the minimum scale just be 1 for iOS family?

aprotyas replied:
```cpp
#if PLATFORM(IOS_FAMILY)
static constexpr double minimumZoomScale = 1;
#else
static constexpr double minimumZoomScale = 0.2;
#endif
```

hortont424 replied:
possible!

hortont424 replied:
ok, making this change too yes

hortont424 committed:
[Unified PDF] Zoom not working in WKWebView
https://bugs.webkit.org/show_bug.cgi?id=288186
rdar://145047812

Reviewed by Abrar Rahman Protyasha.

By default, WebKit respects "user-scalable=no". Safari overrides this to force
pages to always be scalable. For multiple different reasons, all fixed in this patch,
the default viewport configuration for main frame UnifiedPDFPlugin had "user-scalable=no" set.

* Source/WebCore/html/PluginDocument.cpp:
Don't explicitly set user-scalable=no...

* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::ViewportConfiguration::pluginDocumentParameters): Deleted.
Remove this PDFPlugin-specific code from WebCore.

* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::forceAlwaysUserScalableMaximumScale const):
(WebCore::ViewportConfiguration::forceAlwaysUserScalableMinimumScale const):
Don't let "force always user-scalable" to ever *reduce* the available scale range, only increase it.
This prevents it from reducing our PDF maximum scale from 6x to 5x.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::viewportParameters):
Move the UnifiedPDFPlugin default viewport parameters computation to a static method on UnifiedPDFPlugin.
Critically, change the maximum scale to match UnifiedPDFPlugin's normal maximum scale,
and set `allowsUserScaling` to true (e.g. "user-scalable=yes"). The combination
of these two things make scaling work when in the default "do not force always user-scalable" state.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::resetViewportDefaultConfiguration):
Ask UnifiedPDFPlugin for its viewport configuration instead of WebCore.

Canonical link: https://commits.webkit.org/290779@main
